### PR TITLE
PS-7966 Document installing and configuring Percona Server with ZenFS…

### PR DESCRIPTION
**Pull request by patrickbirch**
… from packages (8.0)

modified:   source/index.rst
	modified:   source/installation/apt_repo.rst
	modified:   source/myrocks/install.rst
	modified:   source/myrocks/zenfs.rst
        modified:   source/installation.rst (removed zenfs tarball reference)

### CODE REVIEW

File: doc/source/myrocks/zenfs.rst [hunk 1]
```diff
@@ -13,99 +13,191 @@ a Linux kernel implementation which supports NVMe Zoned Namespaces. The kernel
...
-    Used: 0 MB
-    Reclaimable: 0 MB
-    Space amplification: 0%
+2. Install the `zenfs` package. The Percona Server for MySQL with MyRocks and the ZenFS plugin package is listed in the :ref:`installing_from_binary_tarball` section of the *Percona Server for MySQL* installation instructions. 
+   
+   .. code-block:: bash
+	
+      $ sudo apt install percona-server-server-zenfs
+    
+#. Install the `libzbd` package. A set of instructions are `here <https://ubuntu.pkgs.org/20.04/ubuntu-universe-amd64/libzdb-dev_3.1-0.2ubuntu1_amd64.deb.html>`__.
```
Comment (percona-ysorokin) [inline doc/source/myrocks/zenfs.rst]: This paragraph `Install the libzbd package` can now be completely removed.
`libzbd` is now linked statically into RocksDB Storage Engine for Percona Server, so users don't have to think about this dependency anymore.

File: doc/source/myrocks/zenfs.rst [hunk 2]
```diff
@@ -13,99 +13,191 @@ a Linux kernel implementation which supports NVMe Zoned Namespaces. The kernel
...
+
+#. Initialize ZenFS on ``nvme0n2``.
+
+   .. code-block:: bash
+	
+	  $ sudo -H -u mysql zenfs mkfs --zbd=nvme0n2 --aux_path=/var/lib/mysql_zenfs_aux_ nvme0n2 --finish_threshold=0 --force
+		
+   .. note::
+	
+		Currently, the ``zenfs`` utility does not report an error when the utility does not create an "aux-path" directory because of the permissions. If needed, do the following steps manually.
```
Comment (percona-ysorokin) [inline doc/source/myrocks/zenfs.rst]: @patrickbirch WD implemented a fix for this issue (https://jira.percona.com/browse/WDCZEN-18)
https://github.com/westerndigitalcorporation/zenfs/pull/87
Which means that this is now expected behavior. If we want to configure ZenFS to use a directory inside `/var/lib/` (owned by `root:root` without write permissions for others), we need to do all the steps described below manually
```
sudo mkdir
sudo chown
sudo chmod
```
and only then
```
sudo -H -u mysql zenfs mkfs
```

File: doc/source/myrocks/zenfs.rst [hunk 3]
```diff
@@ -13,99 +13,191 @@ a Linux kernel implementation which supports NVMe Zoned Namespaces. The kernel
...
-    zenfs df --zbd=nvme1n2
-    Free: 7563 MB
-    Used: 0 MB
-    Reclaimable: 0 MB
-    Space amplification: 0%
+2. Install the `zenfs` package. The Percona Server for MySQL with MyRocks and the ZenFS plugin package is listed in the :ref:`installing_from_binary_tarball` section of the *Percona Server for MySQL* installation instructions. 
+   
+   .. code-block:: bash
+	
+      $ sudo apt install percona-server-server-zenfs
```
Comment (percona-ysorokin) [inline doc/source/myrocks/zenfs.rst]: Here and everywhere, there will be no Percona Server packages with `-zenfs` suffix. From now on, `percona-server-server`  includes support for ZenFS on selected platforms automatically.

File: doc/source/myrocks/zenfs.rst [hunk 4]
```diff
@@ -13,99 +13,191 @@ a Linux kernel implementation which supports NVMe Zoned Namespaces. The kernel
...
+
+   .. note::
+
+	  An Ubuntu 20.04 user may be required to build the library from sources if Percona Server for MySQL was built with `-DWITH_ZBD=system`.
+	
+#. Install the RocksDB plugin package. This package copies ``ha_rocksdb.so`` into a predefined location. **The RocksDB storage engine is not enabled**.
+
+   .. code-block:: bash
+	
+	  $ sudo apt install percona-server-rocksdb-zenfs
```
Comment (percona-ysorokin) [inline doc/source/myrocks/zenfs.rst]: Remove `-zenfs` suffix

File: doc/source/myrocks/zenfs.rst [hunk 5]
```diff
@@ -13,99 +13,191 @@ a Linux kernel implementation which supports NVMe Zoned Namespaces. The kernel
...
+	
+	  $ sudo chmod 640 /dev/nvme0n2
+
+#. Change the scheduler for ``nvme0n2`` to ``mq_deadline``.
+
+   .. code-block:: bash
+	
+	  $ sudo bash -c 'echo mq-deadline > /sys/class/block/nvme0n2/queue/scheduler'
+
+#. Create an auxiliary directory for ZenFS, for example, ``/var/lib/mysql_aux``. The ZenFS auxiliary directory is a regular (POSIX) file system used internally to resolve file locks and shared access. There are no strict requirements for the location but the directory must be write accessible for the `mysql:mysql` UNIX system user account. Each ZBD must have an individual auxiliary directory. This directory is at the same level as "/var/lib/mysql", which is the default Percona Server for MySQL directory.
```
Comment (percona-ysorokin) [inline doc/source/myrocks/zenfs.rst]:
```suggestion
#. Create an auxiliary directory for ZenFS, for example, ``/var/lib/mysql_aux``. The ZenFS auxiliary directory is a regular (POSIX) file system directory used internally to resolve file locks and shared access. There are no strict requirements for the location but the directory must be write accessible for the `mysql:mysql` UNIX system user account. Each ZBD must have an individual auxiliary directory. This directory is recommended to be at the same level as "/var/lib/mysql", which is the default Percona Server for MySQL directory.
```

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

File: doc/source/myrocks/zenfs.rst [hunk 6]
```diff
@@ -13,99 +13,191 @@ a Linux kernel implementation which supports NVMe Zoned Namespaces. The kernel
 must be configured with `zone block device support
 enabled <https://zonedstorage.io/linux/config/#kernel-configuration>`__.
```
Comment (LenzGr) [inline doc/source/myrocks/zenfs.rst]: Note that the URL has changed to https://zonedstorage.io/docs/linux/config

File: doc/source/myrocks/zenfs.rst [hunk 7]
```diff
@@ -13,99 +13,191 @@ a Linux kernel implementation which supports NVMe Zoned Namespaces. The kernel
...
-        If you are re-initializing an already initialized ZBD, you must manually clear the ``/var/lib/mysql_aux`` directory and add the additional ``--force`` parameter.
-
-        .. sourcecode:: bash
-
-            zenfs mkfs --zbd=nvme1n2 --aux_path=/var/lib/mysql_aux  --force
-
-
+   The ``<block_device_name>`` can have a short name designation which is the ``<short_block_device_name>``. For the purposes of this document the ``block_device_name`` is ``/dev/nvme0n2`` and the short name is ``nvme0n2``. 
+    
+The primary supported platform is `Debian 11.1 <https://www.debian.org/releases/bullseye/debian-installer/index>`__.
```
Comment (LenzGr) [inline doc/source/myrocks/zenfs.rst]: Let's be precise what we mean by "supported" here - the only platform that we (Percona) provide a binary release for. There are other Linux OSes that now provide full support for ZenFS, but we don't necessarily provide installable packages for these (yet). We should add a pointer to https://zonedstorage.io/docs/distributions/linux so users can educate themselves about the state of support for various Linux distributions.

File: doc/source/myrocks/zenfs.rst (hunk 6 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Fixed.

File: doc/source/myrocks/zenfs.rst (hunk 7 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Done.

File: doc/source/myrocks/zenfs.rst (hunk 3 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Removed -zenfs

File: doc/source/myrocks/zenfs.rst (hunk 1 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst:53]: removed

File: doc/source/myrocks/zenfs.rst (hunk 4 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst:66]: Removed

File: doc/source/myrocks/zenfs.rst (hunk 5 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst:107]: Updated.

File: doc/source/myrocks/zenfs.rst (hunk 2 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: updated.

File: doc/source/myrocks/zenfs.rst [hunk 8]
```diff
@@ -11,101 +11,190 @@ which places files into zones on a raw zoned block device (ZBD) using
...
-.. code-block:: bash
-    
-    $ sudo apt install percona-server-server-zenfs
-    $ sudo apt install percona-server-rocksdb-zenfs
-
-Running the Installation
-----------------------------
-
-The following steps successfully run Percona Server with MySQL on the  ``--rocksdb-fs-uri=zenfs://dev:<short_block_device_name>``:
+The following procedure installs and configures Percona Server for MySQL on the  ``--rocksdb-fs-uri=zenfs://dev:<short_block_device_name>``.
```
Comment (LenzGr) [inline doc/source/myrocks/zenfs.rst]: I find this sentence somewhat confusing. We're not actually **installing** PS on that device, we rather install using the regular package installation workflow, but then configure it to store it's **data** on ZenFS.

Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Fixed.

File: doc/source/myrocks/zenfs.rst [hunk 9]
```diff
@@ -11,101 +11,191 @@ which places files into zones on a raw zoned block device (ZBD) using
...
+
+   .. code-block:: bash
+	
+	  $ sudo chmod 640 /dev/nvme0n2
+
+#. Change the scheduler for ``nvme0n2`` to ``mq_deadline``.
+
+   .. code-block:: bash
+	
+	  $ sudo bash -c 'echo mq-deadline > /sys/class/block/nvme0n2/queue/scheduler'
```
Comment (LenzGr) [inline doc/source/myrocks/zenfs.rst]: Note that this only changes the I/O scheduler at run time, this configuration will be lost upon the next reboot. I wonder if we should explain on how to configure this in a more permanent fashion.

File: doc/source/myrocks/zenfs.rst [hunk 10]
```diff
@@ -11,101 +11,191 @@ which places files into zones on a raw zoned block device (ZBD) using
...
+
+#. Initialize ZenFS on ``nvme0n2``.
+
+   .. code-block:: bash
+	
+	  $ sudo -H -u mysql zenfs mkfs --zbd=nvme0n2 --aux_path=/var/lib/mysql_zenfs_aux_ nvme0n2 --finish_threshold=0 --force
+		
+   .. note::
+	
+		 If you must configure ZenFS to use a directory inside ``/var/lib`` (owned by ``root:root`` without write permissions for other user accounts), do the following steps manually:
```
Comment (LenzGr) [inline doc/source/myrocks/zenfs.rst]: Won't SELinux/AppArmor prevent the process from writing to these arbitrary locations? Do we take this into account?

File: doc/source/myrocks/zenfs.rst [hunk 11]
```diff
@@ -115,11 +205,23 @@ Shut down the server and use the following command to backup a ZenFS file system
 
 .. note::
 
-    At this time, a logical backup is the only backup type supported by the Percona Server MyRocks tables located on ``ZenFS``.
+    At this time, a logical backup is the only backup type supported by the *Percona Server for MySQL*. The MyRocks tables are located on ``ZenFS``.
```
Comment (LenzGr) [inline doc/source/myrocks/zenfs.rst]:
```suggestion
    At this time, a logical backup is the only backup type supported by *Percona Server for MySQL*. The MyRocks tables are located on ``ZenFS``.
```
I struggle with the second sentence. We should include a better explanation here.

File: doc/source/myrocks/zenfs.rst [hunk 12]
```diff
@@ -115,11 +205,23 @@ Shut down the server and use the following command to backup a ZenFS file system
...
+
+			
+Known Limitations
+====================================================================
+
+After a reboot the NVME ZBD configuration ("/dev/nvme02" in our examples) may not reappear. The issue may be OS-dependent and may be managed by the system admins. One or more of the following events may have occurred:
+
+- A reboot changes the active "scheduler" from "[mq-deadline]"
+
+- A reboot resets the device permissions from "640/mysql:mysql" to "660/root:disk".
```
Comment (LenzGr) [inline doc/source/myrocks/zenfs.rst]: Can't we ask our experts on how to make these changes permanently? There must be ways to remove these limitations.

File: doc/source/myrocks/zenfs.rst (hunk 11 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Fixed.

File: doc/source/myrocks/zenfs.rst (hunk 10 above)
Comment (percona-ysorokin) [reply doc/source/myrocks/zenfs.rst]: Yes, @LenzGr you are right. Most probably `/var/lib/mysql_zenfs_aux_<zbd_name>` will be blocked by SELinux/AppArmor. This requires further investigation. At the moment I suggest to add a note
```
You may need to manually update AppArmor/SELinux rules to give mysqld write access to
/var/lib/mysql_zenfs_aux_<zbd_name> 
```

File: doc/source/myrocks/zenfs.rst (hunk 9 above)
Comment (percona-ysorokin) [reply doc/source/myrocks/zenfs.rst]: @LenzGr Here
https://unix.stackexchange.com/questions/375600/how-to-enable-and-use-the-bfq-scheduler
I found something similar but for `bfq` scheduler
```
Create /etc/udev/rules.d/60-scheduler.rules if it did not exist and add:

ACTION=="add|change", KERNEL=="sd*[!0-9]|sr*", ATTR{queue/scheduler}="bfq"
```
This is probably the right way to go

File: doc/source/myrocks/zenfs.rst (hunk 12 above)
Comment (percona-ysorokin) [reply doc/source/myrocks/zenfs.rst]: As @AgileStas suggested, device ownership can also be changed at boot time via `udev`
https://www.thegeekdiary.com/how-to-configure-device-file-owner-group-with-udev-rules/

File: doc/source/myrocks/zenfs.rst (hunk 10 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Updated based on discussion.

File: doc/source/myrocks/zenfs.rst (hunk 9 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Added as note

File: doc/source/myrocks/zenfs.rst (hunk 12 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Updated.

File: doc/source/myrocks/zenfs.rst [hunk 13]
```diff
@@ -1,125 +1,242 @@
...
 
 
-#. You can start Percona Server for MySQL with any of the known methods. For example, add this option to ``my.cnf``:
+   .. code-block:: bash
 
-    .. sourcecode:: text
+      $ sudo apt install gnupg2	
+      $ wget https://repo.percona.com/apt/percona-release_latest.$(lsb_release -sc)_all.deb	
+      $ sudo dpkg -i percona-release_latest.$(lsb_release -sc)_all.deb	
+      $ sudo percona-release setup ps80
```
Comment (LenzGr) [inline doc/source/myrocks/zenfs.rst]: These steps can be slightly simplified (I'm working on updating the generic percona-release docs accordingly):
```suggestion
      $ wget https://repo.percona.com/apt/percona-release_latest.generic_all.deb	
      $ sudo apt install gnupg2 lsb-release ./percona-release_latest.generic_all.deb	
      $ sudo percona-release setup ps80
```

Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Done.

File: doc/source/myrocks/zenfs.rst [hunk 14]
```diff
@@ -1,125 +1,241 @@
...
+The following procedure installs Percona Server for MySQL and then configures 
+``--rocksdb-fs-uri=zenfs://dev:<short_block_device_name>`` for data storage.
 
-        zenfs mkfs --zbd=nvme1n2 --aux_path=/var/lib/mysql_aux
+.. note::
 
-    .. note::
+   The ``<block_device_name>`` can have a short name designation which is the ``<short_block_device_name>``. For the purposes of this document, the ``block_device_name`` is ``/dev/nvme0n2`` and the short name is ``nvme0n2``. 
+    
+For the moment, Percona provides a binary release for `Debian 11.1 <https://www.debian.org/releases/bullseye/debian-installer/index>`__. `Other Linux distributions <https://zonedstorage.io/docs/distributions/linux/>`__ are adding support for ZenFS, but Percona does not provide installation packages for those distributions yet.
```
Comment (percona-ysorokin) [inline doc/source/myrocks/zenfs.rst]: Actually, we build binaries with support for ZenFS for 2 platforms:
1. Debian 11.1 out-out-of-the box (no changes required).
2. Ubuntu 20.04 Focal (users will have to manually install 5.11 HWE kernel patched with CAP_SYS_ADMIN fix)

On Focal, binaries with ZenFS support can still be run on standard 5.4 kernel provided that users do not enable ZenFS functionality (do not specify `--rocksdb-fs-uri` option)

**Review by percona-ysorokin — COMMENTED** [review]
Minor comment regarding potential Ubuntu Focal support

File: doc/source/myrocks/zenfs.rst (hunk 14 above)
Comment (patrickbirch) [reply doc/source/myrocks/zenfs.rst]: Updated.

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request merged**